src/librustc_resolve/diagnostics.rs

```diff
@@ -444,6 +444,52 @@ impl SomeTrait for Foo { // ok!
 ```
 "##,
 
+E0406: r##"
+A function is referring to an associated type which hasn't been declared in
+the trait. Example of erroneous code:
+
+```compile_fail
+trait Foo {
+    type Bar;
+
+    // error: function signature contains a reference to `Self::Baz`, but
+    //        `Baz` hasn't been declared as an associated type of the trait
+    fn return_bool(&self, &Self::Bar, &Self::Baz) -> bool;
+}
```

Comment:
When I try to compile this code, I get E0220, not E0406:

```
$ cat scratch.rs 
trait Foo {
    type Bar;

    fn return_bool(&self, &Self::Bar, &Self::Baz) -> bool;
}

fn main() {}
$ rustc scratch.rs 
scratch.rs:4:40: 4:49 error: associated type `Baz` not found for `Self` [E0220]
scratch.rs:4     fn return_bool(&self, &Self::Bar, &Self::Baz) -> bool;
                                                    ^~~~~~~~~
scratch.rs:4:40: 4:49 help: run `rustc --explain E0220` to see a detailed explanation
error: aborting due to previous error
$ rustc --version
rustc 1.11.0-nightly (5c2a5d449 2016-06-11)
```

I've found that `grep`ing around in the codebase to see where an error code is being generated can provide a clue as to what kind of examples would trigger that error (although usually not a very good clue, given the vastness of my ignorance of the compiler internals); in this case, it looks like E406 will get emitted [during `rustc_resolve::Resolver.resolve_generics`](https://github.com/rust-lang/rust/blob/51d2d3da8/src/librustc_resolve/lib.rs#L1868) (the mapping of error to error-code happens [elsewhere in the file](https://github.com/rust-lang/rust/blob/51d2d3da8/src/librustc_resolve/lib.rs#L232-L234)) ... but I don't know what that means or when it happens.